### ISSUE TITLE
User-defined nonlinear functions for categorical distribution with identity link function

### ISSUE THREAD

**Issue by danheck** [paul-buerkner/brms#301]
As the title says, we would like to combine a user-defined function that generates model predictions for category probabilities with the categorical distribution (with an identity link function). The transformation to the probability space is performed within the function. Currently, the categorical distribution does not allow to specify nonlinear predictions for the category probabilities and does not support the identity link. Is it possible to include this functionality in brms? A simple example of what we would like to achieve is below.

```
# model-implied probabilities for 4 categories
stan_model_prob <- "
vector model_prob(real d, real g) {
  real p_d = Phi(d);
  real p_g = Phi(g);
  vector[4] prob;
  prob[1] = p_d + (1-p_d)*p_g;
  prob[2] = (1-p_d)*(1-p_g);
  prob[3] = (1-p_d)*p_g;
  prob[4] = p_d + (1-p_d)*(1-p_g);
  return prob / 2.0;
}
"

# example data
dat <- data.frame(id = rep(1:30, each = 20),
                  response = sample.int(4, size = 60, replace = TRUE))

# nonlinear model (currently not supported for categorical)
form <- brmsformula(response ~ model_prob(d, g) - 1,
                    d ~ 1 + (1 | id),
                    g ~ 1,
                    family = categorical(link = "identity"),
                    nl = TRUE)
prior <- c(set_prior("normal(0, 1)", class = "b", nlpar = "g"),
           set_prior("normal(0, 1)", class = "b", nlpar = "d"))
make_stancode(form, data = dat, stan_funs = stan_model_prob, prior = prior)
```

**Comment by paul-buerkner** [comment]
Thanks for the suggestion! This indeed sounds like a good idea. I will see if I can make this possible.

**Comment by paul-buerkner** [comment]
There is a conceptual problem we have to think about. Currently, the categories of a categorical model (except for the reference category) are treated as distributional parameters, each having its own predictor term. I can active non-linear models for these parameters (at least I think so), but they will neither share parameters nor can we specify a non-linear function predicting all categories at once. The latter two do not come naturally to the current framework so I am not yet sure if I can make this possible in brms (at least for now).

**Comment by danheck** [comment]
Thank you for looking into this. Would it be an option use an array of vectors for the distributional parameters in combination with `categorical(link="identity")`? Thereby, the predictions can be nonlinear and share parameters.

Below, there is a minimal working example for this. We first generated the model structure using `make_stancode` with `family=gaussian()`. In order to obtain a model for categorical data, we changed four details in the Stan code generated by `make_stancode`. The model can be compiled and fitted with `stan_model`. 

```
library("brms")
model <- "
real model(real d, real g, real cov, int y) {
  return 1.;  // placeholder
}
"
form <- brmsformula(response ~ model(d, g, cov, 1),
                    d ~ cov + (1 | id),
                    g ~ 1,
                    nl = TRUE)
prior <- c(set_prior("normal(0, 1)", class = "b", nlpar = "g"),
           set_prior("normal(0, 1)", class = "b", nlpar = "d"))
dat <- data.frame(id = rep(1:5, each = 10),
                  response = as.integer(1:2),
                  cov = rnorm(50))
make_stancode(form, data = dat, prior = prior, stan_funs = model)
```

Manual changes to the Stan code produced by `make_stancode`:

```
functions{
// replace model predictions
vector model(real d, real g, real cov, int y) {
  real pd = Phi(d);
  real pg = Phi(g);
  vector[2] prob = rep_vector(0., 2);
  prob[1] = pd + (1-pd)*pg;
  prob[2] = (1-pd)*(1-pg);
  return prob;
}
}


data{
...
// vector[N] Y;  // response variable
int<lower=1,upper=2> Y[N];
}


model{
...
// distributional parameters: array of vectors
// vector[N] mu; 
vector[2] mu[N];  // 2 = number of categories

// likelihood
// target += normal_lpdf(Y | mu, sigma);  
for (n in 1:N) {
	target += categorical_lpmf(Y | mu[n]);
}
}
```

After making these changes, the model can be compiled and fitted:

```
brms_data <- make_standata(form, data = dat)
fit0 <- sampling(stan_mpt, data = brms_data)
```

**Comment by paul-buerkner** [comment]
In general, this would be an option, but it doesn't play nicely with all of brms' internals in this regard. The categorical family has always been a stepchild in brms trying to fit in somewhere and as of brms 2.0 has found its home in the distributional regression framework. In the end it doesn't matter if I specify the predictors for each category via separate vectors or a single array of vectors in Stan. Sharing parameter across model parts in brms is currently only possible in very specific situations and I have yet to come up with a general solution that will (and has to) go beyond the categorical family. Otherwise, we will end up with too much special case coding, making it hard to maintain brms.

**Comment by danheck** [comment]
Thank you for the explanation, it is of course more important that the internals of brms are consistent than implementing a special case. And the current example seems to be related to more general cases such as multivariate nonlinear models. 

So for the time being, we can just adapt the brms-stancode manually as shown above, which already allows to use many of the nice features of the brms modeling framework.

**Comment by paul-buerkner** [comment]
Just to elaborate a little bit more. It is not just about not doing any special case coding (there is already quite a bit of special case coding in place to make categorical models work at all), but it the current framework, implementing the proposed feature would require **a lot** more of it.

Multivariate non-linear models are possibls as of brms 2.0, but the main problem of not being able to equate parameters across model parts remains.

**Comment by singmann** [comment]
Should be closed, new issue is #381.
